File: 60_Permutation_Sequence/Solution.py
class Solution:

    def getPermutation(self,n,k):

        jiecheng = {}
        for i in range(1,10):
            t = 1
            for j in range(1,i+1):
                t *= j

            jiecheng[i] = t

        nums = [i for i in range(1,n+1)]
        c = n
        while k > 1:
            if jiecheng[c]>k:
                if jiecheng[c-1] > k:
                    c -= 1
                    continue
                elif jiecheng[c-1] == k:
                    c -= 1

            shit = (k-1) // jiecheng[c-1]

            temp = nums[n-c+shit]
            for i in range(n-c+shit,n-c,-1):
                nums[i] = nums[i-1]
            nums[n-c] = temp
            k = k - jiecheng[c-1]*shit
        result = ''
        for i in nums:
            result += str(i)

        return result

s = Solution()
print(s.getPermutation(3,3,))
print(s.getPermutation(4,9))
print(s.getPermutation(3,2))
print(s.getPermutation(3,4))
print(s.getPermutation(4,7))
print(s.getPermutation(2,2))

File: 60_Permutation_Sequence/Solution.cc

#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

class Solution{
    public:
        static void nextPermutation(vector<int>& nums) {
            int n = nums.size();
            if(n <= 1) return;

            int i=n-2;
            int j=n-1;
            while(i>=0 && nums[i]>=nums[i+1]) i-=1;
            if(i>=0){
                while(nums[j]<=nums[i]) j-=1;
                if(i<j){
                    int t = nums[i];
                    nums[i] = nums[j];
                    nums[j] = t;
                }
            }

            i=i+1, j=n-1;
            while(i<=j){
                int t= nums[i];
                nums[i] = nums[j];
                nums[j] = t;
                i+=1;
                j-=1;
            }
        }

        static string getPermutation(int n, int k){
            vector<int> nums;
            for(int i=1;i<=n;i++) nums.push_back(i);

            int i=1;
            while(i<k){
                nextPermutation(nums);
                i++;
            }
            string ret;
            for(int c:nums) ret+= '0'+c;
            return ret;
        }
};

int main(){
    string line;
    while(getline(cin, line)){
        istringstream is(line);
        int n, k;
        is>>n, is>>k;
        string ret = Solution::getPermutation(n, k);
        cout<<line<<endl;
        cout<<ret<<endl<<endl;
    }
    return 0;
}
